SI-7475 Private members aren't inheritable, findMember overhaul

retronym opened the pull request:
Review by @adriaanm @odersky @xeno-by

Details in the commit messages. The last one poses a few questions
for which I need input.

retronym committed:
SI-7475 Refactor findMember computation into a class

This affords us:

  - a better chance to share code with `findMembers` again
  - the ability to break up the logic into smaller methods for
    scrutability and JIT-tability.

The cost is the use of the heap rather than the stack for the
working area of the computation. But I didn't notice a difference
between `locker.lib` and `quick.lib` after this change.

I've left the old implementation in place with an assertion to
verify identical results. In the next commit, this stepping stone
will be removed, and we'll update `findMembers` to share the code.

Some problem areas have been highlighted and will be addressed
after the refactoring phase is complete.

retronym committed:
SI-7475 findMember and findMembers: estranged no more

Swathes of important logic are duplicated between `findMember`
and `findMembers` after they separated on grounds of irreconcilable
differences about how fast they should run:

    d905558 Variation #10 to optimze findMember
    fcb0c01 Attempt #9 to opimize findMember.
    71d2ceb Attempt #8 to opimize findMember.
    77e5692 Attempty #7 to optimize findMember
    275115e Fixing problem that caused fingerprints to fail in
    e94252e Attemmpt #6 to optimize findMember
    73e61b8 Attempt #5 to optimize findMember.
    04f0b65 Attempt #4 to optimize findMember
    0e3c70f Attempt #3 to optimize findMember
    41f4497 Attempt #2 to optimize findMember
    1a73aa0 Attempt #1 to optimize findMember

This didn't actually bear fruit, and the intervening years have
seen the implementations drift.

Now is the time to reunite them under the banner of `FindMemberBase`.

Each has a separate subclass to customise the behaviour. This is
primarily used by `findMember` to cache member types and to assemble
the resulting list of symbols in an low-allocation manner.

While there I have introduced some polymorphic calls, the call sites
are only bi-morphic, and our typical pattern of compilation involves
far more `findMember` calls, so I expect that JIT will keep the
virtual call cost to an absolute minimum.

Test results have been updated now that `findMembers` correctly
excludes constructors and doesn't inherit privates.

Coming up next: we can actually fix SI-7475!

retronym committed:
SI-7475 Private members are not inheritable

It turns out `findMembers` has been a bit sloppy in recent
years and has returned private members from *anywhere* up the
base class sequence. Access checks usually pick up the slack
and eliminate the unwanted privates.

But, in concert with the "concrete beats abstract" rule in
`findMember`, the following mishap appeared:

    scala> :paste
    // Entering paste mode (ctrl-D to finish)

    trait T { def a: Int }
    trait B { private def a: Int = 0 }
    trait C extends T with B { a }

    // Exiting paste mode, now interpreting.

    <console>:9: error: method a in trait B cannot be accessed in C
           trait C extends T with B { a }
                                      ^

I noticed this when compiling Akka against JDK 8; a new private method
in the bowels of the JDK was enough to break the build!

It turns out that some finesse in needed to interpret SLS 5.2:

> The private modifier can be used with any definition or declaration
> in a template. They are not inherited by subclasses [...]

So, can we simply exclude privates from all but the first base class?
No, as that might be a refinement class! The following must be
allowed:

    trait A { private def foo = 0; trait T { self: A => this.foo } }

This commit:

  - tracks when the walk up the base class sequence passes the
    first non-refinement class, and excludes private members
  - ... except, if we are at a direct parent of a refinement
    class itself
  - Makes a corresponding change to OverridingPairs, to only consider
    private members if they are owned by the `base` Symbol under
    consideration. We don't need to deal with the subtleties of
    refinements there as that code is only used for bona-fide classes.
  - replaces use of `hasTransOwner` when considering whether a
    private[this] symbol is a member.

The last condition was not grounded in the spec at all. The change
is visible in cases like:

    // Old
    scala> trait A { private[this] val x = 0; class B extends A { this.x } }
    <console>:7: error: value x in trait A cannot be accessed in A.this.B
           trait A { private[this] val x = 0; class B extends A { this.x } }
                                                                       ^
    // New
    scala> trait A { private[this] val x = 0; class B extends A { this.x } }
    <console>:8: error: value x is not a member of A.this.B
           trait A { private[this] val x = 0; class B extends A { this.x } }
                                                                       ^
Furthermore, we no longer give a `private[this]` member a free pass
if it is sourced from the very first base class.

    trait Cake extends Slice {
      private[this] val bippy = ()
    }

    trait Slice { self: Cake =>
      bippy // BCS: Cake, Slice, AnyRef, Any
    }

The different handling between `private` and `private[this]`
still seems a bit dubious.

The spec says:

> An different form of qualification is private[this]. A member M
> marked with this modifier can be accessed only from within the
> object in which it is defined. That is, a selection p.M is only
> legal if the prefix is this or O.this, for some class O enclosing
> the reference. In addition, the restrictions for unqualified
> private apply.

This sounds like a question of access, not membership. If so, we
should admit `private[this]` members from parents of refined types
in `FindMember`.

AFAICT, not too much rests on the distinction: do we get a
"no such member", or "member foo inaccessible" error? I welcome
scrutinee of the checkfile of `neg/t7475f.scala` to help put
this last piece into the puzzle.

One more thing: findMember does not have *any* code the corresponds
to the last sentence of:

> SLS 5.2 The modifier can be qualified with an identifier C
> (e.g. private[C]) that must denote a class or package enclosing
> the definition. Members labeled with such a modifier are accessible
> respectively only from code inside the package C or only from code
> inside the class C and its companion module (§5.4).
> Such members are also inherited only from templates inside C.

When I showed Martin this, he suggested it was an error in the
spec, and we should leave the access checking to callers of
that inherited qualified-private member.

retronym commented:
BTW, I'll probably squash the first two commits after review. The first one calls _both_ the old and new `findMember` and asserts identical results, to help convince the beloved reviewers that is a straight refactoring, and the logic changes are restricted to the last two commits.

adriaanm commented:
>  The first one calls both the old and new findMember and asserts identical results, to help convince the beloved reviewers that is a straight refactoring, and the logic changes are restricted to the last two commits.

This reviewer thanks the author kindly, and shall now embark on the tour of the PR.

adriaanm commented:
A real pleasure to review and maintain. Minor comments only.

adriaanm commented:
BTW, I'm okay leaving the commits un-squashed. I could imagine pointing to these commits in the future when asking for reviewer-soothing evidence.

retronym commented:
Thanks for the comments, I will address them tomorrow.

gkossakowski commented on src/reflect/scala/reflect/internal/tpe/FindMembers.scala:
```diff
@@ -0,0 +1,285 @@
+import TypesStats._
+
+trait FindMembers {
+  this: SymbolTable =>
```
It's nitpicking but is the whole SymbolTable needed? Wouldn't it be enough to require `Symbols with Types`?

retronym replied:
We need: `Symbols with Types with Names with Scopes with StdNames`.

Its a judgement call, but I think I'll leave it at `SymbolTable` for now.

It would be good to analyze the opportunity for self-type narrowing across the board. Maybe the best outcome would be to identify commonly used subsets, and represent them with a type. I'm a bit wary of using long lists of explicit dependencies as it adds a burden when moving around from file to file.

retronym replied:
Of course, if you tell us that incremental compilation will massively profit from this discipline, there is more incentive :)

gkossakowski replied:
That's where I was going with this. Fewer dependencies (and less fat ones) are always better but I agree it's probably better to leave for across the board refactoring. I'll be looking into that once 2.11.0-RC1 is out.

retronym committed:
SI-7475 Address review comments in FindMembers

  - comment the cases in `isNewMember`
  - check both sides for privacy (doesn't influence behaviour of
    any tests, but it means that method can be understood without
    understanding the order of traversal of members.)
  - hoist `findAll`
  - break out of the loop in `FindMembers` as soon as we determine
    that the candidate member is matched by an already-found member
    and can be discarded.

retronym commented:
A recent bug fix to master (#3357) exposed a missing piece of this puzzle. I've rebased and amended the second last commit.

``` patch
git diff 487a54a
diff --git a/src/compiler/scala/tools/nsc/transform/OverridingPairs.scala b/src/compiler/scala/tools/nsc/transform/Ove
index 870eafb..bbd11ef 100644
--- a/src/compiler/scala/tools/nsc/transform/OverridingPairs.scala
+++ b/src/compiler/scala/tools/nsc/transform/OverridingPairs.scala
@@ -24,7 +24,12 @@ abstract class OverridingPairs extends SymbolPairs {
     /** Symbols to exclude: Here these are constructors and private/artifact symbols,
      *  including bridges. But it may be refined in subclasses.
      */
-    override protected def exclude(sym: Symbol) = sym.isPrivateLocal || sym.isArtifact || sym.isConstructor
+    override protected def exclude(sym: Symbol) = (
+         sym.isPrivateLocal
+      || sym.isArtifact
+      || sym.isConstructor
+      || (sym.isPrivate && sym.owner != base) // Privates aren't inherited. Needed for pos/t7475a.scala
+    )

     /** Types always match. Term symbols match if their member types
      *  relative to `self` match.
```

Adding to the commit comment:

```
 Makes a corresponding change to OverridingPairs, to only consider
    private members if they are owned by the `base` Symbol under
    consideration. We don't need to deal with the subtleties of
    refinements there as that code is only used for bona-fide classes.
```

Re-review by @adriaanm

adriaanm commented on src/compiler/scala/tools/nsc/transform/OverridingPairs.scala:
```diff
@@ -24,7 +24,12 @@ abstract class OverridingPairs extends SymbolPairs {
+         sym.isPrivateLocal
+      || sym.isArtifact
+      || sym.isConstructor
+      || (sym.isPrivate && sym.owner != base) // Privates aren't inherited. Needed for pos/t7475a.scala
```
I haven't been able to come up with an example, but I'm surprised we needn't look at privateWithin / accessBoundary.

adriaanm commented:
Ugh, looks like the conflict is in src/reflect/scala/reflect/runtime/JavaUniverseForce.scala -- code generated by run/t6240-universe-code-gen.scala. /cc @xeno-by

adriaanm commented:
Rebasing.

retronym commented:
PLS REBASE --UPDATE_CHECK @ADRIAANM

By which I mean, thanks!

adriaanm commented:
pleasure!

PS: I updated test/partest a while ago to do an ant build.pack, so rebase --update-check can be implemented as g rebase -i scala/master --exec "rm -rf build/pack ; test/partest --update-check  <tests touched by this commit>", except that the exec lines should come before the "edit" lines in this case

retronym commented:
The not-a-member vs member-but-inaccessible distinction is only really observing that, in the latter case, you can always move the call site _somewhere_ to make it accessible:

```
scala> trait T { private def foo = 0; private[this] def bar = 0 }; class C { self: T => this.foo; this.bar }
<console>:7: error: method foo in trait T cannot be accessed in C with T
       trait T { private def foo = 0; private[this] def bar = 0 }; class C { self: T => this.foo; this.bar }
                                                                                             ^
<console>:7: error: value bar is not a member of C with T
       trait T { private def foo = 0; private[this] def bar = 0 }; class C { self: T => this.foo; this.bar }
                                                                                                       ^

scala> trait T { private def foo = 0; private[this] def bar = 0; class C { self: T => this.foo; this.bar } }
<console>:7: error: value bar is not a member of T.this.C with T
       trait T { private def foo = 0; private[this] def bar = 0; class C { self: T => this.foo; this.bar } }
                                                                                                     ^
```

retronym commented:
The difference between `private[this]` and `private` here is the residual issue that needs to be cleaned up, it isn't supported by the spec.

With that, and https://github.com/scala/scala-dist/pull/130/files, we should at least have `private` under control.
